src/NuGet.Core/NuGet.Protocol/Providers/DownloadResourceV3Provider.cs

```diff
@@ -33,7 +33,14 @@ public DownloadResourceV3Provider()
 
                 // If index.json contains a flat container resource use that to directly
                 // construct package download urls.
-                var packageBaseAddress = serviceIndex.GetServiceEntryUri(ServiceTypes.PackageBaseAddress)?.AbsoluteUri;
+                var endpoint = serviceIndex.GetServiceEntryUri(ServiceTypes.PackageBaseAddress);
+                var packageBaseAddress = endpoint?.AbsoluteUri;
+
+                // Check if the source is not HTTPS.
+                if (endpoint.Scheme == Uri.UriSchemeHttp && endpoint.Scheme != Uri.UriSchemeHttps && !source.PackageSource.AllowInsecureConnections)
+                {
+                    throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, Strings.Error_HttpServiceIndexUsage, source.PackageSource.SourceUri, endpoint));
```

Comment:
These checks are being performed in every resource provider, which has the risk that we accidentally miss a provider, or in the future someone adds a new provider and doesn't add this AllowInsecureConnections check.

If instead we check only in the service index provider, when we load the service index, then I think it can reduce risk of human errors.  However, it would have the side-effect that if a custom nuget feed has a custom resource in the service index that NuGet doesn't use, we might throw exceptions. But it's a tradeoff that you might find acceptable.

Reply:
Great idea!